### ISSUE TITLE
TODO items from tests.Rraw

### ISSUE THREAD

**Issue by MichaelChirico** [Rdatatable/data.table#2572]
Items cut out from the end of `tests.Rraw` when this issue was first created:

- [x] Tests involving GForce functions needs to be run with optimisation level 1 and 2, so that both functions are tested all the time.
- [x] Add test for fixed bug #5519 - dcast returned error when a package imported data.table, but dint happen when "depends" on data.table. This is fixed (commit 1263 v1.9.3), but not sure how to add test. [#2571]
- [x] test and highlight in docs that negatives are fine and fast in forderv (ref R wish #15644)
- [x] bring ?setkey description up to date; #3399
- [x] tests of freading classes like Date and the verbose messages there. [#2545]
- [x] Test mid read bump of logical to character, rereads original True and False character strings faithfully.
- [x] add examples of multiple LHS (name and position) and multiple RHS to example(":=")
- [x] tests on double in add hoc by
- [x] test on -i that retain key e.g.  DT[-4] and DT[-4,sum(v),by=b] should both retain key
- [x] test on out of bound i subsets e.g. 6:10 when DT has 7 rows, and mixed negative and positive i integer is error.
- [x] test that ordered subsets when i is unkeyed now retain x's key (using is.sorted(f__))
- [x] add FAQ that  eval() is evaled in calling frame so don't need a, then update SO question of 14 March. See the test using variable name same as column name. Actually, is that true?  Need "..J".
- [x] revisit [SO answer](https://stackoverflow.com/a/10230625/3576984). Can there be a better way than the repeated `lapply(f,eval,.SD)`?
- [x] change all 1 to 1L internally (done in data.table.R, other .R to do)
- [x] check the "j is named list could be inefficient" message from verbose than Chris N showed recently to 15 May
- [x] !make sure explicitly that unnamed lists are being executed by dogroups!
- [x] Add to warning about a previous copy that class<-, levels<- can also copy whole vector.  *Any* fun<- form basically.
- [x] use looped := vs set test in example(":=") or example(setnames) to test overhead in [.data.table is tested to stay low in future.
- [x] add tests on smaller examples with NAs for 'frankv', even though can't compare to base::rank.

---

`TODO` items sprinkled throughout the rest of the file:

 - [ ] confirm with Matt the behavior of `rbindlist` w.r.t. naming of input in edge cases -- rbindlist should look for the first non-empty data.table - New changes (from Arun). Explanation below: Even if data.table is empty, as long as there are column names, they should be considered. Ex: What if all data.tables are empty? What'll be the column name then? If there are no names, then the first non-empty set of names will be allocated. I think this is the way to do it...
 - [ ] "Ragged" files that should have `fill = TRUE` generate warnings, not errors; as such, we need another way to trigger a STOP() inside fread.c. options(warn=2) isn't enough.
 - [ ] Add tests of `NaN` behavior in `forder` (testing internal behavior since can't compare to `base::order`)
 - [x] Think of and add more tests for `rbindlist`
 - [x] `allow.cartesian` is ignored if `jsub[1L]` has `:=`; could still warn if `i` has duplicates
 - [x] Add `comment=="#"`.   Ensure only after last field is observed.
 - [x] Add `integer64` support for `transpose`
 - [ ] Add a test when `fill=FALSE`, but `blank.lines.skip=TRUE`, when the same effect as in tests 1585 should happen
 - [x] Fix and add test for cases like this: `a,b,c\n1,2,3\n4,5,6\n7,8,9,6\n1,2,3\n...` -- extra column in line 4, but we've only detected 3
 - [x] Add tests for `nomatch=NA` in non-equi-join suite (tests 1640-ish)... tested, but takes quite some time... so commented for now _[split off as #3411]_
 - [x] 0.0 written as 0, but TODO #2398, probably related to the 2 lines after l==0 missing coverage in writeFloat64
 - [x] Add `strict.numeric` (default `FALSE`) to `all.equal.data.table()`? _[split off as #3410]_
 - [ ] Begin to deprecate `logicalAsInt` argument
 - [x] #1416 TODO (trailing tabs on most but not at the beginning and a "-" intended to mean missing but taken as text column name)
 - [ ] Begin deprecation of `..x` as valid column names

---

`TO DO` items sprinkled throughout the rest of the file:

 - [ ] the eval() enabled you to use the 'a' in the calling scope, not 'a' in the TESTDT.  TO DO: document this.
 - [ ] Test that changing colnames keep key in sync. TO DO: will have to do this for secondary keys, too, when implemented.
 - [ ] Test shallow copy verbose message from := adding a column, and (TO DO) only when X is NAMED.
 - [ ] Used to be warning about invalid .internal.selfref detected and fixed. As from v1.8.3 data.table() returns a NAMED==0 object, and key<- appears not to copy that. But within functions, key<- would still copy. TO DO: add tests....
- [ ] ### ----> Therefore this TO DO may not be necessary here anymore (added by Arun 26th Jan 2014) ---> # TO DO when options(datatable.pedantic=TRUE): test(680.5, rbindlist(l[c(2,6)]), warning="Column 1 of item 2 is type 'double', inconsistent with column 1 of item 1's type ('integer')")
 - [ ] Test non ascii characters when passed as character by, #2134 TO DO: reinstate. Temporarily removed to pass CRAN's Mac using C locale (R-Forge's Mac is ok)
 - [ ] TO DO: add warning about inefficiency when datatable.pedantic=TRUE
 - [ ] TO DO: why doesn't last group have x=d, maybe groups=i in dogroups
 - [ ] the i groups when no match don't get carried through (would be hard to implement this and very unlikely to be useful. Just break into compound query, if needed to be used in j, to get them to carry through. TO DO: add to FAQ.
 - [ ] A is needed otherwise read as double with loss of precision (TO DO: should detect and bump to STR)
 - [ ] \" protected ok, but \ needs taking off too (TO DO)
 - [ ] Takes too long for routine use. TO DO: move to a long running stress test script
 - [ ] Selection of columns, copy column to maintain the same as R <= 3.0.2, in Rdevel, for now; otherwise e.g. setkey changes the original columns too.  TO DO: could allow shallow copy, perhaps.
 - [ ] TO DO: roll to -Inf.  Also remove -Inf and test rolling to NaN and NA
 - [ ] rounding of milliseconds, workaround, TO DO: #5445
 - [ ] TO DO, by checking for balanced embedded quotes
 - [ ] to do: add tests for :=

**Comment by MichaelChirico** [comment]
My idea for testing:

> use looped := vs set test in example(":=") or example(setnames) to test overhead in [.data.table is tested to stay low in future.

is to run:

```
t1 = system.time({
    DT = data.table(a = 1:10)
    for (jj in paste0(seq_len(100))) DT[ , (jj) := 10:1]
})

t2 = system.time({
    DT = data.table(a = 1:10)
    for (jj in paste0(seq_len(100))) set(DT, , jj, 10:1)
})

ratio = T1[3L]/t2[3L]
```

Then error if `ratio` is outside `(1, R]`. On my machine right now, the ratio is about 8; I guess setting `R` to 10 would be too tight/potentially machine-dependent. How about choosing `R = 50`? `25`?

**Comment by MichaelChirico** [comment]
@mattdowle the following doesn't appear to be true anymore? unless I missed something about what's intended...

> test that ordered subsets when `i` is unkeyed now retain `x`'s key (using `is.sorted(f__)`)

```
DT1 = data.table(ID = 1:4, v = 4:1, key = 'ID')
DT2 = data.table(ID = 3:6, w = c(3, 4, 1, 2))
key(DT1[DT2])
# NULL
key(DT1[DT2, on = 'ID'])
# NULL
DT1[DT2]
#    ID  v w
# 1:  3  2 3
# 2:  4  1 4
# 3:  5 NA 1
# 4:  6 NA 2
```

FWIW, I see that the [line `is.sorted(f__)` has been commented out](https://github.com/Rdatatable/data.table/blob/master/R/data.table.R#L870).

**Comment by MichaelChirico** [comment]
@st-pasha you're more familiar with the `fread` test suite, is this covered by now?

> Test mid read bump of logical T/F to character, collapse back to T and F.

**Comment by st-pasha** [comment]
I don't remember if there is such a test specifically, but it's easy to add:
```
```

**Comment by MichaelChirico** [comment]
@st-pasha thanks, I wasn't sure if there's an exact row number to pick to ensure re-read.

Tried beefing it up, but `fread` doesn't take `"T"` `"F"` to be `bool`?

```
DT = data.table(A=rep("True", 2200), B="FALSE", C="T", D='0')
DT[111, LETTERS[1:4] := .("here", "be", "dra", "gons")]
fwrite(DT, f<-tempfile())
fread(f, verbose=TRUE)
```

**Comment by MichaelChirico** [comment]
@mattdowle this one appears to be obsolete as well?

> check the "j is named list could be inefficient" message from verbose than Chris N showed recently to 15 May

`grep "named list R/*"` comes up blank.

**Comment by MichaelChirico** [comment]
re:

> tests on double in add hoc by

I am making a test along the lines of:

```
DT = data.table(a = c(1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2), v = 1:12,
                jig = 10^(289:300))
DT[ , sum(v), by = .(a_jiggle = a + jig * .Machine$double.xmin)]
#     a_jiggle V1
#  1:        1  6
#  2:        1  4
#  3:        1  5
#  4:        1  6
#  5:        2  7
#  6:        2  8
#  7:        2  9
#  8:        2 10
#  9:        2 11
# 10:        2 12
```

I seem to recall it being documented somewhere what exactly the internal limits of accuracy are/how to query this for `data.table`; on my machine, it appears that roughly disturbances less than `1e-17` don't make a difference... not sure the robust/machine-independent way to design this test, however.

**Comment by MichaelChirico** [comment]
> * use looped := vs set test in example(":=") or example(setnames) to test overhead in [.data.table is tested to stay low in future.

@jangorecki this can't make it to the CRAN test suite, but do we have anywhere we could readily drop in such a test (more like a dashboard, really) in dev?

**Comment by MichaelChirico** [comment]
> test and highlight in docs that negatives are fine and fast in forderv (ref R wish #15644)

`?forder` already mentions the use of `-` in both `setorder` and `setorderv`. If anything the documentation can be cleaned up -- hardly relevant to be pointing out differences of 1.9.2 vs 1.9.4 (i.e. 4.5-5 years ago)... that set `?setorder` could go for a tidying

**Comment by MichaelChirico** [comment]
> check the "j is named list could be inefficient" message from verbose than Chris N showed recently to 15 May

`grep "named list" R/*` returns nothing; I think this is obsolete

**Comment by MichaelChirico** [comment]
> tests on double in add hoc by

IIUC we're not really supposed to be testing things like numerical accuracy. Limitations on double sorting are spelled out in the documentation IIRC... and anyway we haven't seen any issues raised on this in a long time. So I think this is covered.

**Comment by jangorecki** [comment]
> > * use looped := vs set test in example(":=") or example(setnames) to test overhead in [.data.table is tested to stay low in future.
> 
> @jangorecki this can't make it to the CRAN test suite, but do we have anywhere we could readily drop in such a test (more like a dashboard, really) in dev?

There is benchmark.Rraw, or if there is no, there is a plan to have one. This will be proper place for all tests that needs to call `system.time()`.

**Comment by MichaelChirico** [comment]
FWIW, did the following benchmark of `set` vs. `:=`:

```
library(microbenchmark)

set.seed(5419681)

BB = 100L
times = CJ(rows = 10^(3:6), cols = c(10, 50, 100, 1000), 
           type = c(':=', 'set'), repetition = 1:BB)

for (ii in seq_len(nrow(times))) {
  times[ii, duration := switch(type, ':=' = {
    DT = data.table(ID = seq_len(rows))
    t0 = get_nanotime()
    DT[ , paste0('V', 1:cols) := lapply(integer(cols), function(...) rnorm(.N))]
    get_nanotime() - t0
  }, 'set' = {
    DT = data.table(ID = seq_len(rows))
    t0 = get_nanotime()
    for (jj in 1:cols) set(DT, , paste0('V', jj), rnorm(rows))
    get_nanotime() - t0
  })]
}

times[ , .(ratio = duration[type == ':=']/duration[type == 'set'],
           avg = mean(duration)), keyby = .(rows, cols, repetition)
       ][ , as.list(c(absolute_seconds = mean(avg)/1e9, summary(ratio))), 
          keyby = .(rows, cols)]
#      rows cols absolute_seconds      Min.   1st Qu.    Median      Mean   3rd Qu.     Max.
#  1: 1e+03   10      0.001741932 0.7370478 1.1379052 1.3440595 1.6102717 1.8700619 6.837264
#  2: 1e+03   50      0.006072508 0.2342039 1.0315886 1.0851785 1.1190119 1.1960915 3.100473
#  3: 1e+03  100      0.011430631 0.7444117 0.9521297 0.9939637 1.0238520 1.0270249 2.828019
#  4: 1e+03 1000      0.130151601 0.0758904 0.8265715 0.8483162 0.8603380 0.8788563 1.589417
#  5: 1e+04   10      0.011169461 0.8837293 0.9913126 1.0319755 1.0625403 1.0509171 4.792800
#  6: 1e+04   50      0.055652297 0.4357905 0.9764087 1.0051055 1.0014451 1.0414789 1.100584
#  7: 1e+04  100      0.113367405 0.5040770 0.9475957 0.9983624 1.0157946 1.0521188 1.922424
#  8: 1e+04 1000      0.828476190 0.4775837 0.9758829 1.0105570 1.0339757 1.0569433 2.278120
#  9: 1e+05   10      0.108609403 0.5402728 0.9431956 0.9554793 0.9494869 0.9684458 1.060773
# 10: 1e+05   50      0.468080595 0.2467847 0.9949998 1.0055778 1.0566599 1.0153622 3.914609
# 11: 1e+05  100      0.835524829 0.6005924 0.9898779 0.9985906 1.0174447 1.0120674 1.760639
# 12: 1e+05 1000      8.048499634 0.6872025 0.9424409 0.9633926 0.9518192 0.9829113 1.153556
# 13: 1e+06   10      0.850079092 0.8575973 0.9832558 1.0349987 1.0446556 1.1356624 1.238097
# 14: 1e+06   50      4.069398375 0.9998059 1.0349430 1.0514445 1.0533430 1.0711666 1.170846
# 15: 1e+06  100      8.082150041 0.9584103 1.0200243 1.0371611 1.0534353 1.0468116 2.466068
# 16: 1e+06 1000     89.837404857 0.8318387 1.2215667 1.2447751 1.2805784 1.3248808 1.637098
cat('\nTotal Time:', times[ , sum(duration)/1e9/60], 'minutes')
# Total Time: 378.1927 minutes
```
Could go for doing this plain overnight as I was doing all sorts of other stuff while this was going on (including moving the computer about with me)

**Comment by MichaelChirico** [comment]
> Add to warning about a previous copy that class<-, levels<- can also copy whole vector.  _Any_ fun<- form basically.

Not sure where it is the intention to put this. Axing barring that.

**Comment by jangorecki** [comment]
why assign `rnorm`? it costs
good to call `gc` before each timing start.

**Comment by MichaelChirico** [comment]
> why assign `rnorm`? it costs

best to just assign `TRUE`?

will do re:`gc`

**Comment by MichaelChirico** [comment]
updated set/:= benchmark, 20x as fast when not using `rnorm`

**Comment by jangorecki** [comment]
Results presented above are difficult to interpret. I would find that useful to melt results into something like:
```
rows, cols, first_set, first_:=, mean_set, mean_:=, median_set, rmedian_:=, relative_first, relative_mean, relative_median
```

**Comment by MichaelChirico** [comment]
> !make sure explicitly that unnamed lists are being executed by dogroups!

honestly no idea what this is about. IINM `dogroups` only entryway in R is when computing groups and `!GForce` [here](https://github.com/Rdatatable/data.table/blob/master/R/data.table.R#L1834).

Pretty trivial to see that there's plenty of coverage of this in tests already (e.g. some places in #3406 where optimization is turned off, among probably hundreds of others).

Alternative interpretation would be that `.ok` should be `FALSE` if `j` is a named list? Though this is obviously not the case (maybe it was at some point?)

Gonna mark this one done

**Comment by MichaelChirico** [comment]
> tests of freading classes like Date and the verbose messages there.

Marking this complete since it's covered in a separate PR/issue already

**Comment by MichaelChirico** [comment]
> confirm with Matt the behavior of `rbindlist` w.r.t. naming of input in edge cases -- rbindlist should look for the first non-empty data.table - New changes (from Arun). Explanation below: Even if data.table is empty, as long as there are column names, they should be considered. Ex: What if all data.tables are empty? What'll be the column name then? If there are no names, then the first non-empty set of names will be allocated. I think this is the way to do it...

@mattdowle any feedback here?

**Comment by MichaelChirico** [comment]
> * "Ragged" files that should have `fill = TRUE` generate warnings, not errors; as such, we need another way to trigger a STOP() inside fread.c. options(warn=2) isn't enough.

@mattdowle ditto here... could you split off into a new issue if this is still a concern?

**Comment by MichaelChirico** [comment]
> # TO DO: TODO: think of and add more tests for rbindlist

This is quite vague and 5 years old... closing as duplicate of the [40+ `rbindlist` issues outstnding](https://github.com/Rdatatable/data.table/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+rbindlist)

**Comment by MichaelChirico** [comment]
> `allow.cartesian` is ignored if `jsub[1L]` has `:=`; could still warn if `i` has duplicates

AFAICT this is subsumed by #3077; closing here

**Comment by MichaelChirico** [comment]
> * Add `comment=="#"`.   Ensure only after last field is observed.

Presumably this is just #856; closing here

**Comment by MichaelChirico** [comment]
> Add `integer64` support for `transpose`

This is a full issue; filed as #3408 and closing here

**Comment by MichaelChirico** [comment]
> Add a test when `fill=FALSE`, but `blank.lines.skip=TRUE`, when the same effect should happen

Unclear what "same effect" is referring to; I see tests 1584.x that seem to cover this case? @arunsrinivasan this is under you per `git blame` (admittedly 3+ years ago), any idea?

**Comment by MichaelChirico** [comment]
> Fix and add test for cases like this: `a,b,c\n1,2,3\n4,5,6\n7,8,9,6\n1,2,3\n...` -- extra column in line 4, but we've only detected 3

This seems to be handled already:

```
s = "a,b,c
1,2,3
4,5,6
7,8,9,6
1,2,3"

fread(s)
#        a     b     c
#    <int> <int> <int>
# 1:     1     2     3
# 2:     4     5     6
```

> Warning message:
> In fread(s) :
  Stopped early on line 4. Expected 3 fields but found 4. Consider fill=TRUE and comment.char=. First discarded non-empty line: <<7,8,9,6>>

```
fread(s, fill=TRUE)
#        a     b     c    V4
#    <int> <int> <int> <int>
# 1:     1     2     3    NA
# 2:     4     5     6    NA
# 3:     7     8     9     6
# 4:     1     2     3    NA
```

**Comment by MichaelChirico** [comment]
> Add tests for `nomatch=NA` in non-equi-join suite (tests 1640-ish)... tested, but takes quite some time... so commenting for now

@jangorecki this set of tests could be added to the GitLab suite? 

Is the `TODO` covered by the commented tests?

```
# TODO: add tests for nomatch=NA..
# tested, but takes quite some time.. so commenting for now
# nqjoin_test(x, y, 3L,1643.0)
# nqjoin_test(dt1,dt2,3L,1652.0)

# nqjoin_test(  x,dt2,1L,1644.0) # without NA only in x
# nqjoin_test(  x,dt2,2L,1645.0)
# nqjoin_test(  x,dt2,3L,1646.0)
# nqjoin_test(dt1,  y,1L,1647.0) # without NA only in i
# nqjoin_test(dt1,  y,2L,1648.0)
# nqjoin_test(dt1,  y,3L,1649.0)
```

**Comment by MichaelChirico** [comment]
> 0.0 written as 0, but TODO #2398, probably related to the 2 lines after l==0 missing coverage in writeFloat64

As stated, this is just #2398

**Comment by MichaelChirico** [comment]
> Add `strict.numeric` (default `FALSE`) to `all.equal.data.table()`?

@jangorecki can revisit this and file as a separate issue if still viable? But I guess no because of the difficulty in testing `identical` for `numeric` across platforms? Or could include it but discourage it "use at your own risk"?

**Comment by MichaelChirico** [comment]
> Begin to deprecate `logicalAsInt` argument

This was 1.11.0 (May 2018); can begin the deprecation process with next release?

**Comment by MichaelChirico** [comment]
> `#1416 TODO (trailing tabs on most but not at the beginning and a "-" intended to mean missing but taken as text column name)`

As stated, this is just #1416

**Comment by MichaelChirico** [comment]
> Begin deprecation of `..x` as valid column names

This is also 1.11.0 (May 2018). But see #3321, making things more complicated...

**Comment by MichaelChirico** [comment]
Same as #678 -- better to handle this with a linter
